### PR TITLE
Tidy advanced gui panel

### PR DESCRIPTION
**Pull request by feerrenrut**
### Summary of the issue:
With the introduction of the advanced panel in the settings dialog #9200, there were a couple of UI issues, the text did not wrap properly, and the warning message was only announced when the dialog received focus (not when moving from the category list to the panel itself).
Several times during development, old settings dialog windows were not being destroyed, this meant that the settings dialog could not be opened a second time. The check for multiple instances was trying to serve two purposes:
- Stop users from opening a second settings dialog while one was already open, changing settings from two places at once could have unintended and confusing outcomes.
- Warn developers that dialogs are not closing / being disposed of correctly.

### Description of how this pull request fixes the issue:
The long text is now wrapped at the width of the panel. The "Warning!" text is larger and bold.

The warning text is announced when the panel receives focus.

A state is appended to the settings dialog instances, this keeps track of whether the dialog has been destroyed and it is removed when the dialog is deleted. We can then give the standard "user friendly" error when the settings dialog is already open, and log an error when the window is destroyed but not deleted.

### Testing performed:
The settings dialog should be reported as prior versions.

The panel description should be reported when the dialog gets focus and the focused control is:
- a control on the document formatting panel
- a control on the advanced panel
This description should only be reported once, regardless of whether the change in focus is due to alt+tab (swapping applications), ctrl+tab (swapping settings categories), or moving from the category list to the settings panel

### Known issues with pull request:
- Due to the delays of garbage collection, if the NVDA settings dialog is closed and opened again quickly, an error may be added to the log.

#### Alternatives:
- The text could be used as a label for a checkbox which enables / disables the rest of the panel. The checkbox must be checked before modifications can be made. Perhaps it's state is saved in the configuration?

### Change log entry:

Section: New features, Changes, Bug fixes

Changes:
```
 - Descriptive text is now announced for the advanced settings panel and the document formatting panel (#9239).
```

### CODE REVIEW

**Comment by michaelDCurran** [comment]
I think it should be a warning checkbox at the beginning of the Advanced panel that remembers its state for the lifetime of NVDA, but is not saved in the config. This is similar to Firefox.

**Comment by leonardder** [comment]
I second this idea, however, I wouldn't put the warning on the checkbox itself if possible.

If I"m correct, you can do the following:

1. Create another static box (grouping) called warning
2. Add a static text with the warning to that grouping. I recall that magic will automatically assign that static text to accDescription on the grouping, but that's the weak part in my idea, I'm not entirely sure.
3. You can then add a check box with a relatively short label.

**Comment by DrSooom** [comment]
> • The text could be used as a label for a checkbox which enables / disables the rest of the panel. The checkbox must be checked before modifications can be made. Perhaps it's state is saved in the configuration?

This … no, a checkbox alone, which value is stored in the "nvda.ini" doesn't solve this GUI issue at all, but a button called "Understood" like in Firefox when entering "about:config". All settings in this panel are unaccessible (= write protection) unless this button has pressed.

Well, the checkbox for automatically pressing this button (= disables this button and makes all other settings writable) can be added in this panel too. But we should prevent the normal user for changing something in this panel without showing him the warning first. So a combination of a button and a checkbox should be the best way to solve this GUI issue.

**Comment by feerrenrut** [comment]
So after trying lots of options with little success, I think the only viable option is to use the AccPropServer to:
- override the description for the panel
- set the panel role to property page

I am then setting the name of the panel using `self.SetName`, because overriding it with AccPropServer did not result in it being spoken!

One remaining issue is that the alt+tabbing to the dialog results in `getDialogText` running (in behaviours.py ) which essentially makes the dialog speak twice. For now I have resolved this by removing `controlTypes.ROLE_PROPERTYPAGE` from the collection of child roles included. We will need a better solution for this.

This result is that we can have these bits of "dialog explaining text" announced.
See the updated description for full test results.

**Comment by leonardder** [comment]
Is this considered ready for review, or are you still considering changing some things?

**Comment by leonardder** [comment]
Nevertheless, I have tested the code. Here are some thoughts you might want to consider:

1. Removing event_gainFocus from the NVDA appModule broke the configuration profile changes.
2. Have you considered subclassing wx.StaticBox in a way that implements an AccPropServer for accDescription on the StaticBox, instead of setting the description on the panel? This requires setting the "warning" label as a label on the StaticBox, instead of adding a separate static text that contains the label. The advantage of this is that the warning is read again when pressing shift+tab from the UI Automation group. It is not read when entering the property page from the ok button, though.
3. Using ROLE_PROPERTYPAGE for settings panels is a nice improvement. Make sure that you do not add an accPropServer on the VoiceSettingsPanel and BrailleSettingsSubPanel though. One way to accomplish this is: only add a server on a panel when it has a title.
4. Rather than giving every settings panel its own AccPropServer instance, you might want to consider registering one accPropServer on MultiCategorySettingsDialog.container, making that a property page and providing the accDescription based on the currently active settings panel. When switching categories, you will have to call SetLabel on the container. Hopefully that will trigger event_namechange on it.

**Comment by feerrenrut** [comment]
> Is this considered ready for review

To start with I wanted to get some feedback on whether this actually improves things, find out what issues there might be from a UX consideration. If we can agree on what is possible, and that it improves the UX then we can get it ready for an actual review.

> broke the configuration profile changes.
Ah, yes. I'll take a look at that.

> One way to accomplish this is: only add a server on a panel when it has a title.

Good idea. 

> you might want to consider registering one `accPropServer` on `MultiCategorySettingsDialog.container`

Yes, probably a good idea. Though, do you know of any disadvantages to having multiple accPropServers registered?

> Hopefully that will trigger event_namechange on it.

I was hoping to be able to remove stuff from the NVDA app module. It might not be possible yet.

**Comment by leonardder** [comment]
> Though, do you know of any disadvantages to having multiple accPropServers registered?

Not really, except for the fact that systems tend to break when an accPropServer isn't properly unregistered, which could possibly happen if NVDA crashes or when one of the panels refuses to load due to an error within the panel's class. While this is not likely to happen in stable versions, the breakage seems to be pretty major sometimes (#8916).

> > Hopefully that will trigger event_namechange on it.
> 
> I was hoping to be able to remove stuff from the NVDA app module. It might not be possible yet.

I'm pretty sure that gainFocus can be limited to do profile name changes only. I recall that I tested setting the current category's name on the container and then stopped with that because the name wasn't announced automatically when I changed categories, but that probably was because of the panel not having the role of property page.

**Comment by feerrenrut** [comment]
I've pushed a few fixes the concerns raised. I also found how to override the getDialogText behavior. This was causing the dialog to be spoken (up to first focused child).  Now the dialog is speaking the current profile as the description. However, this information is also in the title. I am looking for a way to get the accessible name for the dialog to be spoken instead of the title. Another possibility, might be to change the getDialogText behavior so that when encountering a focused property page with a description the recursion stops there.

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: What I've just noticed is that the NVDAObjects.behaviors.Dialog overlay is also applied to property pages, so it looks like you can safely remove the accDescription override for panels like this and only override the accRole. I have commented out this line and it did not result in different behavior.

Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: Hmm, yeah, you are right. The hard question is what to do about the dialog description. Either we need a way to stop the getDialogText on these functions, or a way to override the dialog title.

**Comment by feerrenrut** [comment]
I haven't worked out why yet, but it seems that inspect.exe will give the name that I set via the acc prop server. However NVDA dev info gives the window title.

File: source/gui/settingsDialogs.py
Comment (leonardder) [reply source/gui/settingsDialogs.py]: What part of the dialog description do you mean in this case?

I really like the setup where the config profile is pronounced as part of the dialog description.

How about doing something like this:

1. Only have NVDA Settings in the dialog title.
2. Add a statictext with "editing {profilename}". Then, add the categories list and all other controls. This will make the profile name the first child on the dialog, which will be automatically announced using getDialogText. You no longer have to override accDescription for the dialog
3. When the current profile name changes, update the particular static text with the new profile name. Do something like this on the dialog:
`winUser.user32.NotifyWinEvent(winUser.EVENT_OBJECT_DESCRIPTIONCHANGE, settingsDialog.Handle,winUser.OBJID_CLIENT,winUser.CHILDID_SELF)`
    On the appModule, the only thing you need to do is override event_descriptionChange on the dialog to speak even when the dialog isn't the focus object.

This has the advantage that the visual behavior and non visually perceived behavior match much more than before.

Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: The difficult part with what you are proposing is finding a way to make the profile text look right within the dialog. It's not obvious how to achieve that.

> What part of the dialog description do you mean in this case?

I meant, after alt tab to the dialog. NVDA walks through to the focused control, speaking all static text until it gets there.  I now realise that as long as `PROPID_ACC_DESCRIPTION` is set to something other than white space this will not happen.

I like the idea of using the accessible description for the dialog to announce the profile name. Next I'm going to look into why NVDA does not seem to be getting the `PROPID_ACC_NAME` for the dialog. Inspect.exe shows it, but its not in NVDA's dev info.

**Comment by feerrenrut** [comment]
I have been looking into why I can't seem to override the title with the acc prop server. It seems that there is a failure in comtypes. The strange thing is, this does not affect @michaelDCurran. It is also strange that the same approach works for description. Both of these values are observable from inspect.exe. I have tried from source, and from the PR build. Stepping through the code that fails does not seem to give any hints about what is wrong. It could be that there is a bug with getting the properties on my machine.

Could some others please test the behaviour of [the PR build](https://ci.appveyor.com/api/buildjobs/i8pb0l9r68en46cd/artifacts/output%2Fnvda_snapshot_pr9239-16750%2C58502b47.exe) and report what is announced when using alt+tab to return to the nvda settings dialog? Please also include your windows version, and your python version

**Comment by dnz3d4c** [comment]
When you press alt + tab to switch to the NVDA settings window, NVDA says:
NVDA Settings  dialog

- Windows version: Windows 10 1809(OS build 17763.253) - language Korean
- Python version: 2.7.15

**Comment by michaelDCurran** [comment]
For some reason NVDA is not announcing the correct states for the warning checkbox in the Advanced category when it is toggled. If I go to the Advanced panel and focus the checkbox, it first says it is unchecked which is good. But if I press space, NVDA says nothing, yet the checkbox checks. If I press space again, NVDA announces 'checked' yet the checkbox becomes unchecked. I think there is some kind of delay between the statechange event and wx actually changing the state. Perhaps the event is taking too long to enable all the controls? Can you reproduce this @feerrenrut ?

**Comment by michaelDCurran** [comment]
Sorry if I have reported this previously, but I don't see the comment here.
1. Open an NVDA settings panel (e.g. press NVDA+control+g)
2. Close the dialog with escape.
3. Repeat step 1.
NVDA complains that a settings dialog is already open.

I think there might be another circular reference introduced by the accPropServer on the dialog.
I know it uses a weakref, but perhaps it is to do with the binding of the windowDestroy event, I'm not sure.
Some how the control might be being kept around in a Python frame somewhere on the accPropServer code.

**Comment by leonardder** [comment]
> When you press alt + tab to switch to the NVDA settings window, NVDA says:
> NVDA Settings dialog

Works here as well

* Windows version: Windows 10 1809(OS build 17763.292) - language Dutch
* Python version: 2.7.15

Note that the dialog has two objects in the object tree. The dialog object has NVDA Settings as name. It's parent is a window root object and has NVDA Settings: Normal configuration as its name. This is expected though.

> For some reason NVDA is not announcing the correct states for the warning checkbox in the Advanced category when it is toggled.

I can reproduce this.

**Comment by feerrenrut** [comment]
Thanks for testing @dnz3d4c and @leonardder, was the description announced too? I expected you to get "NVDA Settings dialog normal configuration"
I should have included my windows version: 

So far we have:
- NVDA reports "NVDA Settings: normal configuration  dialog  normal configuration"
  - @feerrenrut : Windows 10 1803 (OS Build 17134.523)
- NVDA reports "NVDA Settings dialog"
  - @dnz3d4c Windows 10 1809(OS build 17763.253)  
  - @leonardder Windows 10 1809(OS build 17763.292)  NVDA reports- language Dutch

> NVDA complains that a settings dialog is already open.

I'll look into this.

> NVDA is not announcing the correct states for the warning checkbox in the Advanced category when it is toggled
 
I'll also look at this.

**Comment by leonardder** [comment]
You could consider updating the comtypes submodule to tag 1.1.7 to see whether it fixes anything.

**Comment by dnz3d4c** [comment]
No, description is not announced

> 
> 
> Thanks for testing @dnz3d4c and @leonardder, was the description announced too? I expected you to get "NVDA Settings dialog normal configuration"
> I should have included my windows version:
> 
> So far we have:
> 
>     * "NVDA Settings: normal configuration  dialog  normal configuration"
>       
>       * @feerrenrut : Windows 10 1803 (OS Build 17134.523)
> 
>     * "NVDA Settings dialog"
>       
>       * @dnz3d4c Windows 10 1809 (OS build 17763.253)
>       * @leonardder Windows 10 1809 (OS build 17763.292)  language Dutch
> 
>     * "NVDA Settings dialog normal configuration"
>       
>       * @michaelDCurran Windows 10 1809 (OS build 18312.0)
> 
> 
> > NVDA complains that a settings dialog is already open.
> 
> I'll look into this.
> 
> > NVDA is not announcing the correct states for the warning checkbox in the Advanced category when it is toggled
> 
> I'll also look at this.

**Comment by leonardder** [comment]
For me, the description was announced properly when alt+tabbing into the dialog.

**Comment by lukaszgo1** [comment]
For me on polish Windows 7 x64 when allt tabbing into the dialog NVDA announces NVDA settings dialog normal configuration.

**Comment by dnz3d4c** [comment]
Oh, it is my mistake. when allt tabbing into the dialog NVDA announces NVDA settings dialog normal configuration.

> 
> 
> No, description is not announced
> 
> > Thanks for testing @dnz3d4c and @leonardder, was the description announced too? I expected you to get "NVDA Settings dialog normal configuration"
> > I should have included my windows version:
> > So far we have:
> > ```
> > * "NVDA Settings: normal configuration  dialog  normal configuration"
> >   
> >   * @feerrenrut : Windows 10 1803 (OS Build 17134.523)
> > 
> > * "NVDA Settings dialog"
> >   
> >   * @dnz3d4c Windows 10 1809 (OS build 17763.253)
> >   * @leonardder Windows 10 1809 (OS build 17763.292)  language Dutch
> > 
> > * "NVDA Settings dialog normal configuration"
> >   
> >   * @michaelDCurran Windows 10 1809 (OS build 18312.0)
> > ```
> > > NVDA complains that a settings dialog is already open.
> > 
> > 
> > I'll look into this.
> > > NVDA is not announcing the correct states for the warning checkbox in the Advanced category when it is toggled
> > 
> > 
> > I'll also look at this.

**Comment by lukaszgo1** [comment]
@feerrenrut I've tested on my English Windows 10 1803 vm without any version of Python installed, and I am getting "NVDA Settings: normal configuration dialog normal configuration" like in your case, so it is something specific to Windows 10 1803. I have all versions of Win 10 either on a physical pcs, or on a vms, so if testing on older builds would help just let me know.

**Comment by feerrenrut** [comment]
Thanks @lukaszgo1 it would be very helpful to know if this has ever worked in windows 10, and when it stopped working. Then if we also know when it started working again we can try to handle that version range specifically. I had updates install overnight, I get the same behaviour with: 1803 (OS Build 17134.590)

I will also test out a newer release of the comtypes submodule (as suggested by @leonardder) to see if that helps.

**Comment by feerrenrut** [comment]
I tested with comtypes 1.1.7 without any change in behaviour.

While inspecting the change set I noticed a comment that indicated that SetLabel needs to happen after SetupScrolling, which seems strange, however after trying this I am now getting the expected behaviour. Hopefully this works for others also, but bear in mind there are a couple of bugs still present:
- After opening the settings dialog and closing it, you may get an error that the settings dialog is already open. In this case you must restart NVDA to be able to open it again.
- NVDA is not announcing the correct states for the warning checkbox in the Advanced category when it is toggled.

If people are willing to test the [try build](https://ci.appveyor.com/api/buildjobs/s2py0exhfprb8qaq/artifacts/output%2Fnvda_snapshot_try-tidyAdvancedGuiPanel-16770%2C20ab8096.exe) and let me know the behaviour (and their windows version), that would be excellent.

**Comment by lukaszgo1** [comment]
On  Polish Windows 7 x64 I am now getting "NVDA Settings: normal configuration dialog normal configuration" so this unfortunately regressed. Will test on Windows 10 builds soon.

**Comment by leonardder** [comment]
For me, this works as expected with regard to announcing the title "NVDA Settings dialog normal configuration". The window root object is now also covered.

File: source/gui/settingsDialogs.py
Comment (leonardder) [reply source/gui/settingsDialogs.py]: > The difficult part with what you are proposing is finding a way to make the profile text look right within the dialog. It's not obvious how to achieve that.

Isn't it possible to position it centered just below the window title?

Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: It's possible, I'll have to play with it to see how it looks. Because it's not editable it might look a little strange. I have been thinking it makes sense to be able to edit profiles that are not actually active, so maybe it should be a control at some stage.

**Comment by feerrenrut** [comment]
I have updated this, fixing the checkbox issue and the multiple instance error when trying to reopen the settings dialog.

I believe (hope) that I have made the title reporting work on all systems now. Expected behaviour on all systems tested,:
- Windows 7 x64 Version 6.1 (Build 7601: Service Pack 1)
- Windows 10 Version 1803 (OS Build 17134.590)
- Windows 8 Version 6.2 (Build 9200)

A [try build](https://ci.appveyor.com/api/buildjobs/13m9h93mn7u2f8by/artifacts/output%2Fnvda_snapshot_try-tidyAdvancedGuiPanel-16805%2C207d653b.exe) is available to test this on your own machine.

**Review by michaelDCurran — COMMENTED** [review]
I can still reproduce the issue with the warning checkbox. It says nothing the first time I press space, then it is speaking the opposite each time after that. I.e. NVDA is receiving the event before the states have fully changed.

**Comment by leonardder** [comment]
I can confirm @michaelDCurran's finding regarding the checkbox' state not being reflected properly.

I've started JAWS and Narrator alongside NVDA, and they both pick up the state change correctly.

**Comment by leonardder** [comment]
Additional findings:

1. It turns out that only speech fails with regard to announcing the state of the checkbox. Braille follows ok.
2. Removing the restriction from event_stateChange that only announces state changes for focus object, apart from expectedly introducing lots of garbage announcements, does not announce the proper state either.

**Comment by feerrenrut** [comment]
Hmm, you are right. I am getting this behaviour now also. I don't know what is causing this, when I tested this yesterday it was behaving as expected. I'll have to see if I can reproduce the behaviour I got then and look at the differences. It seems a bit strange!

**Comment by michaelDCurran** [comment]
I'm not too sure why this would affect it, but enabling those controls 
is probably quite costly.

If you could queue the enabling of those controls to the next core 
cycle. Say with wx.CallAfter or wx.Calllater, then hopefully the 
checkbox toggling will be nice and fast and the state will be set 
correctly. Then the costly control enabling/disabling can be done after 
that.

**Comment by feerrenrut** [comment]
Before I put the controls in a sub panel, I tried enabling the controls based on type to see if there was one particular type that was the causing the problem. I concluded it was either the `CustomCheckListBox` or the `SelectOnFocusSpinCtrl`. I can confirm this is still the case, removing one, the other, or both gives the expected behaviour.

**Comment by feerrenrut** [comment]
> If you could queue the enabling of those controls to the next core cycle. Say with wx.CallAfter or wx.Calllater

I have actually tried that, with no change in behaviour. I also tried using the UI_UPDATE event.

**Review by LeonarddeR — COMMENTED** [review]
> I concluded it was either the `CustomCheckListBox` or the `SelectOnFocusSpinCtrl`.

These are both custom controls, so may be there's an issue in one of them rather than in the logic we're using to disable/enable the panel?

**Comment by feerrenrut** [comment]
So I'm quite sure its a problem caused by a delay. I added some logging, and could see that when using `wx.callAfter` the call to enable the controls was still happening before NVDA was speaking the checkbox state. I tried using `core.callLater`, and increasing the delay until the check state is spoken, which gives the intended behaviour. However, I think this is something to look at in NVDA, given Narrator speaks the states correctly.

**Comment by feerrenrut** [comment]
I'm wondering if this could be caused by the order that the events are received. Could NVDA be receiving an event to alert it to the state being changed, before the changed state is accessible in the API? EG NVDA is processing the event before WX is. Perhaps this is influenced by our controls somehow? I wonder this, because the speech I get when the item gets focus is correct. Pressing space the first time, nothing is spoken, because NVDA doesn't see a change in state, because it hasn't been updated yet. Pressing space the second time and again, if NVDA gets the event first, it might see the state change from the first press of space, report it (which is the opposite of the actual state). 

Another note, the developer info for the checkbox has the expected state.

**Comment by michaelDCurran** [comment]
I think this simply is a weirdness caused by wx doing something a little 
odd such as firing the event before it updates the control, and also the 
fact that NvDA's GUI and core are in the same thread.


NVDA recievs a focus winEvent for the checkbox. It fetches the MSAA 
object for that event, sets it as the focus in NVDA and also caches the 
current states (accState).

Then the user presses space and we recive a stateChange winEvent. We see 
that the winEvent matches the focused object, so we route it there and 
call event_stateChange on the focus.

In event_stateChange we fetch states (accState) but what we get back is 
not yet updated.


I agree it would be good to know the true simptom for this, but a 
suitable delay will fix it, and we sort of need to either get this in 
(or not) so we can release the first beta :)

**Comment by lukaszgo1** [comment]
Sorry for the delay. I've tested with the try build linked above on:
- Windows 7 x64
- Windows 8 x64
- Windows 10 build 10240
- Windows 10 build 14393
- Windows 10 build 17134
- Windows 10 build 17763
- Windows 10 build 18342 

and on all of them the dialog title is announced correctly.

**Comment by michaelDCurran** [comment]
And the description I would hope? I.e. it said "NVDA settings dialog 
normal configuration"?

As opposed to trying to read all the text from inside the dialog.

**Comment by lukaszgo1** [comment]
@michaelDCurran  wrote:
> 
> And the description I would hope? I.e. it said "NVDA settings dialog normal configuration"? As opposed to trying to read all the text from inside the dialog.


Yes

**Comment by lukaszgo1** [comment]
While the announcement is correct when alt+tabbing to the dialog when using NVDA+t to check the title bar it is always NVDA settings. In the  current Alpha it contains name of the profile in use + the selected category.

**Comment by michaelDCurran** [comment]
As much as this is a loss of information when pressing NVDA+tab, I think 
it was deliberate. The category name is now on the property page, which 
was previously not spoken before.

I think the current behaviour in this pr is more suitable.

**Comment by lukaszgo1** [comment]
What do you mean by saying 
>As much as this is a loss of information when pressing NVDA+tab,

I was talking about NVDA +t. Currently to check which profile we are  editing or in which category of the settings we are it is enough to check the title. bar.  With the try build from this PR it is nessesary to alt+tab from and alt+tab back to the dialog. Adding to this the fact that when alt+tabbing to any of the settings dialog your position is lost I believe that it isn't an acceptable solution.

**Comment by michaelDCurran** [comment]
Sorry, I meant NVDA+t. However, I do take your point that the category 
cannot be read on demand anymore.


But, I'm not entirely sure of the use case where you would be in the 
settings dialog and not know the category. either you alt+tabbed in 
there and it spoke as part of the focus ancestry as focus was inside the 
property page, or you had set the category from the list yourself.

I'm not convinced this problem is enough to block this change?

**Comment by lukaszgo1** [comment]
I don't know if it is an showstopper or not. I am merely pointing out that this change makes retrieving information about currently editing profile and the current category more difficult than it was before.

**Comment by feerrenrut** [comment]
Thanks for testing @lukaszgo1, and you raise a good point. Not something that had been considered yet. We would like to get this merged in, since it's holding back the release. For now we will go with the current approach.  Please see https://github.com/nvaccess/nvda/issues/9321 which could be an approach to resolving this.

**Comment by lukaszgo1** [comment]
don't believe that #9321 would improve this situation. beccause there is no way to reread this control on demand like the title bar.
What about merging this as is and the for the next release simply make the content of title bar like it was before this pr?

**Comment by feerrenrut** [comment]
> there is no way to reread this control on demand like the title bar.

I figured if there was an actual control, focus can be moved back to the control to read it. Obviously this is not as quick as just using `NVDA+t`. It should be noted that when the profile being edited changes, it is still reported. Perhaps a script to report the current profile, those who depend on this set a gesture?

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: I'm not sure whether I already expressed my concerns with this approach. I'm afraid that this way of forcefully destroying instances will obfuscate cases where a dialog isn't destroyed properly due to bugs, such as #8372.

**Comment by lukaszgo1** [comment]
I don't think that not having the active config profile and setting category dialog in the title bar is a big regression however it was nice to have and was intuitive. Introducing additional gesture to reread active profile is in my view an overkill. Is it really technically impossible to have the title bar to contain these information? If it is then simply leaving it as is seems to be the best solution.

File: source/gui/settingsDialogs.py
Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: I do share this concern, there are more complicated things we could do perhaps but I don't want to go down that path right now. This approach will help to stop issues that require the user to restart before they can re-open the settings dialog. Whereas removing it, might make it easier to catch circular references. Essentially there are two purposes. Rather than blocking users from opening the settings dialog, I would prefer a log error in the case of a circular reference detection.

**Comment by feerrenrut** [comment]
> Is it really technically impossible to have the title bar to contain these information?

It's not technically impossible, but then there are other problems that would need to be solved. Right now, we are using the description of the dialog to report the configuration profile. If this is empty, then NVDA will try to describe the dialog automatically. On panels like the advanced panel or the document formatting panel, this results in the panel description being read twice. This happens because the panel has a description that matches the static text that describes the dialog, this is so the description is read when the panel gets focus. NVDA's dialog description logic reads all static text starting at the dialog title, descending down until it gets to the first active control. 

Adding a command to report the active profile is easy. I have a draft pull request: #9325

**Comment by feerrenrut** [comment]
Actually, while testing this again I have noticed that `setLabel` is also overwriting the visible title.

**Comment by leonardder** [comment]
I guess that for now, we should leave the title as is in current Alpha, 
i.e. remove the title changes from this pr.

**Comment by leonardder** [comment]
> NVDA's dialog description logic reads all static text starting at the dialog title, descending down until it gets to the first active control.

I'm wondering whether this behaviour of getDialogText is right. If there is a property page in the dialog that has a description retrieved by getDialogText, and that property page has a focusable (not focused) descendant, than the description of the property page shouldn't end up in the description of the dialog at all, since the user can have it read by focusing a decendant of the property page.

Alternatively, there is some logic in this code that stops adding grouping descriptions to the dialog text. Overall, it is pretty complex and error prone.

**Comment by feerrenrut** [comment]
I think for now we will be best to rely on the NVDA appmodule to override this behaviour. I believe this is ready for review.

File: source/gui/settingsDialogs.py
Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: I think I have addressed this concern now. I have updated the PR description to discuss this.

File: source/appModules/nvda.py
Comment (leonardder) [inline source/appModules/nvda.py]: using controlId is a bit ambiguous here, since it is in fact a windowHandle and not a windowControlID

File: source/gui/nvdaControls.py
Comment (leonardder) [inline source/gui/nvdaControls.py]: Could this cause one of these circular dependency issues?

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: This should be enough

```suggestion
		if self in SettingsDialog._instances:
```

Also makes it py3 compatible :)

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: I leave it up to others to make a final decision, but I think that other software tends to place the restore defaults after the controls. I think I would prefer that here as well, but I don't insist.

File: source/gui/settingsDialogs.py
Comment (leonardder) [inline source/gui/settingsDialogs.py]: We tend to make these instance methods in other places instead of a local function, but I do not care much. Just wanted to point this out.

**Review by LeonarddeR — APPROVED** [review]
Some cosmetic comments that aren't really important and shuldn't block this.

File: source/appModules/nvda.py
Comment (michaelDCurran) [reply source/appModules/nvda.py]: Agreed.

File: source/gui/settingsDialogs.py
Comment (michaelDCurran) [reply source/gui/settingsDialogs.py]: I would like this as well.

**Review by michaelDCurran — APPROVED** [review]

File: source/appModules/nvda.py
Comment (feerrenrut) [reply source/appModules/nvda.py]: Yes, this was a mistake on my part. Good catch.

File: source/gui/nvdaControls.py
Comment (feerrenrut) [reply source/gui/nvdaControls.py]: This represents the ownership and lifetime of the server most accurately. Without this, the relationship and the lifetime of the server is really not clear. It's true that the server holds a weak reference to the control, and something in `SetHwndPropServer` can stop NVDA from being able to exit unless `ClearHwndProps` is called, this is managed by the server binding to the window destroy event.

File: source/gui/settingsDialogs.py
Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: I agree, I didn't start with that because it doesn't seem straight forward to achieve. I also think that there is some benefit to having proximity to the warning message. I don't think that the button should be disabled when the check box is checked, since restoring to defaults should be a safe operation. Perhaps for now, we can leave it as it is, and soon look at how we can provide a more general "restore" function for all panels.

Comment (feerrenrut) [reply source/gui/settingsDialogs.py]: I'm going to merge this as is, if we think it's a big enough problem I will look into how we can adjust this.

Comment (leonardder) [reply source/gui/settingsDialogs.py]: I'm ok with that.

**Review by LeonarddeR — APPROVED** [review]

**Pull request merged**
